### ISSUE TITLE
await Promise<T>.Deferred unexpectedly releasing

### ISSUE THREAD

**Issue by drew-512** [timcassell/ProtoPromise#458]
Hi Tim,

I'm getting a consistent issue that I can't explain -- I'm waiting on a deferred promise and the *second* time I do so, it immediately returns.   

The below client code logging into a server via tcp, where the reply message eventually calls `SetStatus()` causing the deferred promise to release the `await` you see in `DoLogin()`.

This issue was present when I was using 2.9 (or so) then upgraded to 3.1.0 and the issue was unchanged. 

```C#

        public async Promise<Err>           DoLogin() {
            ...
            var err = await Host.AwaitMetaOp(...); // first call -- blocks as expected
            if (err != null) 
                return err;
            ...
            err = await Host.AwaitMetaOp(...);  // ISSUE -- immediately releases!
            if (err != null) 
                return err;
            ...
        }

        public enum OpStatus {
              NotStarted = 0,
              Syncing,
              Busy,
              Synced,
              Closed,
        }

        public Promise<Err>                 AwaitMetaOp(TagSpec attrSpec, IMessage elemVal, ClientOp op) { 
            ...
            op.SetStatus(OpStatus.Syncing);
            var promise = op.Await(OpStatus.Synced);
            _outbox.Enqueue(tx);
            _outboxCount.Release();
            return promise;
        }

        ConcurrentQueue<TxMsg>              _outbox      = new ConcurrentQueue<TxMsg>();
        SemaphoreSlim                       _outboxCount = new SemaphoreSlim(0);

        public Promise<Err>                 Await(OpStatus status) {
            var waiter = new StatusAwaiter(status);
            _waiters ??= new List<StatusAwaiter>(1);
            _waiters.Add(waiter);
            return waiter.Deferred.Promise;
        }
        List<StatusAwaiter> _waiters;


        public bool                         SetStatus(OpStatus newStatus) {
            for (int i = 0; i < _waiters.Count; i++) {
                var wi = _waiters[i];
                if (wi.WaitingFor == newStatus) {
                    _waiters.RemoveAt(i);
                    i--;
                    wi.Deferred.Resolve(_lastErr);
                }
            }
            return true;
        } 

        readonly struct StatusAwaiter {
            public readonly OpStatus              WaitingFor;
            public readonly Promise<Err>.Deferred Deferred;
            
            public StatusAwaiter(OpStatus status) {
                this.WaitingFor = status;
                this.Deferred   = Promise.NewDeferred<Err>(); 
            }
        }
```

**Comment by drew-512** [comment]
If I break when it's issuing the *second* deferred promise, internally I see both the deferred id and promise ID both at 1 -- which I would expect to be 2?  I tried stepping in but it I was hard to know what is the correct behavior since the first deferred promise has already been resolved by this point.

![Screenshot 2024-07-01 at 6 44 28 PM](https://github.com/timcassell/ProtoPromise/assets/35852900/63db7ba5-f806-465c-b19d-d0cef95b3f2c)

**Comment by drew-512** [comment]
Open to feedback on code as well if I'm missing something.

**Comment by timcassell** [comment]
How do you know it continued immediately? Your debug info says the promise is pending.

**Comment by timcassell** [comment]
I think I understand what's happening. I'll post a detailed explanation and suggested fix soon.

**Comment by timcassell** [comment]
The simple explanation is promise continuations are executed synchronously by default. When you call `wi.Deferred.Resolve(_lastErr);` inside `SetStatus` method, it continues the `DoLogin` async method, which recursively calls into `AwaitMetaOp` which calls into `Await` which calls `_waiters.Add(waiter);`. When the stack unwinds back to `SetStatus`, the next time it goes around the loop the `_waiters` list has a new item added to it, so it also processes that. If you check the stacktrace when you hit that breakpoint, you will see it.

You only want each server response to continue a single promise. The simple fix is to simply break out of the loop when the first eligible waiter is found.

```diff
        public bool                         SetStatus(OpStatus newStatus) {
            for (int i = 0; i < _waiters.Count; i++) {
                var wi = _waiters[i];
                if (wi.WaitingFor == newStatus) {
                    _waiters.RemoveAt(i);
-                   i--;
                    wi.Deferred.Resolve(_lastErr);
+                   break;
                }
            }
            return true;
        } 
```

But, this is just a bandaid. Your architecture can only safely handle a single request at a time. 2 or more concurrent requests introduces all sorts of nasty race conditions. (The `ConcurrentQueue` does not guarantee that the response is mapped to the correct request.)

You should instead re-architect your network infrastructure to include a request id in the payload. The response payload should include the same id so you can map it to the correct promise. Something like this:

```cs
public class Network
{
    private const int MaxConcurrentRequests = 4;

    private int _id;
    private readonly ConcurrentDictionary<int, Promise<Response>.Deferred> _deferreds = new ConcurrentDictionary<int, Promise<Response>.Deferred>();
    private readonly AsyncSemaphore _sem = new AsyncSemaphore(MaxConcurrentRequests);

    public async Promise<Response> SendRequest(Request request)
    {
        using (await _sem.EnterScopeAsync())
        {
            int requestId = Interlocked.Increment(ref _id);
            var deferred = Promise<Response>.Deferred.New();
            if (!_deferreds.TryAdd(requestId, deferred))
            {
                _deferreds.TryRemove(requestId, out var oldDeferred);
                oldDeferred.Reject("Response never received.");
                _deferreds.TryAdd(requestId, deferred);
            }

            // Serialize request and requestId and send it to your socket here.

            return await deferred.Promise;
        }
    }

    // Called from your socket after deserializing the requestId and response.
    private void Resond(int requestId, Response response)
    {
        if (!_deferreds.TryRemove(requestId, out var deferred))
        {
            throw new ArgumentException($"No request is waiting for requestId {requestId}.");
        }
        deferred.Resolve(response);
    }
}
```

**Comment by drew-512** [comment]
Appreciate the explanation.  

Yeah, I thought `.Deferred.Resolve()` would release on context relinquish, not synchronously yikes.  Too much time in Go lately it seems.

Basically, on the client side, the idea is to dispatch any number of client requests ("ops"), but as an op's status is altered from a reply via `SetStatus()`, those waiting on those other states are released.   This way I can block whatever easily until the op is at the desired state.  But `.Deferred.Resolve()` having reentrant-ish behavior doesn't work as you've well pointed out.

So the behavior I'm looking for is something that resolves a deferred promise from "outside" the main run loop.  That seems available by moving matching `waiters` into a closure that is executed later -- does anything more efficient come to mind?

**Comment by timcassell** [comment]
> Yeah, I thought `.Deferred.Resolve()` releases on context relinquish, not synchronously yikes. Too much time in Go lately.
> ...
> The behavior I'm looking for is something that resolves a deferred promise from "outside" the main run loop. That seems available by moving matching `waiters` into a closure that is executed later.
> 
> Trying to think of an efficient pattern for that.

You can use `promise.WaitAsync(SynchronizationContext.Current, forceAsync: true)` to make the continuation execute asynchronously on the current context.

> Basically, on the client side, the idea is to dispatch any number of client requests ("ops"), but as an op's status is altered from a reply via `SetStatus()`, those waiting on the op are released. This way I can block whatever easily until the op is at the desired state. However, `.Deferred.Resolve()` having sub-dispatch (vs queued-dispatch) behavior makes that totally wrong as you covered.
> 
> Your mock code is what I had initially but then others calls need to block until the request comes back -- and those other calls can materialize after the request is sent -- hence the introduction of `Await(OpStatus status)`

I assume the "op" is the `StatusAwaiter`. You could use an `AsyncLazy<T>` inside the `StatusAwaiter` which can be awaited multiple times. Or you could try using `Promise<T>.Retainer` if you want to avoid the extra allocation. The mapped responses can (and should imo) be used even with your re-usable "ops".

**Comment by drew-512** [comment]
> I assume the "op" is the `ClientOp`. You could use an `AsyncLazy<T>` inside the `ClientOp` which can be awaited multiple times. Or you could try using `Promise<T>.Retainer` if you want to avoid the extra allocation. The mapped responses can (and should imo) be used even with your re-usable "ops".

Yeah, I already do the mapped responses so this is about deferring others waiting on a particular response state.  

I'm trying to understand what the `AsyncLazy<T>` and `Promise<T>.Retainer` patterns look like since `promise.WaitAsync(SynchronizationContext.Current, forceAsync: true)` seems to address the behavior where one wants deferred invocation rather than immediate invocation.

**Comment by timcassell** [comment]
> this is about deferring others waiting on a particular response state

I see. I thought you were just trying to wait until completion. If simply deferring the continuation solves your issue, then feel free to ignore the rest of my comments, as I do not have the full context of your architecture.

> I'm trying to understand what the `AsyncLazy<T>` and `Promise<T>.Retainer` patterns look like

For this case you would need an `AsyncLazy` for each state. But that would probably complicate your code rather than simplify it, so your list scanning approach is probably good enough (as long as you are only operating on that unsynchronized list on 1 thread).
